go/pkg/cas/upload.go

```diff
@@ -965,6 +1106,67 @@ func marshalledRequestSize(d *repb.Digest) int64 {
 	return marshalledFieldSize(reqSize)
 }
 
+type missingDir struct {
+	relPath  string
+	item     *uploadItem
+	digested *digested
+}
+
+// missingDirs returns missing directories on the node path from the root to
+// each of the path specified in the tree.
+// For example, if the tree contains only "foo/bar" and "foo/baz", then
+// missingDirs returns "foo" and the root ".".
+//
+// All tree keys must be clean relative paths.
+func missingDirs(tree map[string]*digested, absRoot string) []missingDir {
```

Comment:
should we call this `partialMerkleTree`? I was a bit confused as I expected `missingDirs` to somehow go over/return the directories _not_ listed in `tree` (eg `foo` has `bar/` and `baz/`, `tree` include `bar/`, this returns `baz/`).

Reply:
In an earlier version, this func was called `partialMerkleTree` but then I realized that it doesn't quite build merkle tree. Rather, it builds those missing ancestors.

I agree with your point and renamed it to `missingAncestors`. I think the reader should no longer expect children to be returned. WDYT?

Reply:
with the UploadInput refactoring, this func now mutates `tree`, and the caller actually receives the merkle tree, so done.